nntrainer/tensor/memory_pool.cpp

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file   memory_pool.cpp
+ * @date   11 August 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is Memory Pool Class
+ */
+
+#include <numeric>
+#include <vector>
+
+#include <memory_pool.h>
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+
+/**
+ * @brief Request Memory from memory pool
+ * @note start_time is inclusive, but end_time is exclusive
+ */
+unsigned int MemoryPool::requestMemory(size_t bytes, unsigned int start_time,
+                                       unsigned int end_time) {
+  if (mem_pool != nullptr)
+    throw std::invalid_argument(
+      "Deallocate memory pool before requesting more memory");
+
+  if (end_time <= start_time)
+    throw std::invalid_argument(
+      "Invalid validity range for the requested memory");
+
+  memory_size.push_back(bytes);
+  memory_validity.push_back({start_time, end_time});
+
+  /** invalidate min_pool_size if already there */
+  min_pool_size = 0;
+
+  return memory_size.size() - 1;
+}
+
+/**
+ * @brief Planner the layout with memory planner
+ *
+ * @details The efficiency of the planner is calculated as the ratio of the
+ * theoretical minimum memory requirement divided by the memory requirement
+ * given by the memory planner.
+ *
+ * @details planLayout can be called multiple times as this does not perform
+ * any allocation but rather just plans the layout and stores the layout.
+ * Subsequent call to this function will overwrite any existing layout.
+ */
+double MemoryPool::planLayout(const MemoryPlanner &planner) {
+  if (mem_pool != nullptr)
+    /** mem_pool must be deallocated when planLayout is being called */
+    throw std::runtime_error("Planning memory layout after allocation");
+
+  if (memory_size.empty())
+    throw std::runtime_error("Planning memory layout for empty pool");
+
+  /** calculate min_pool_size if not already calculated */
+  if (min_pool_size == 0)
+    min_pool_size = calcMinMemoryRequirement();
+
+  pool_size = planner.planLayout(memory_size, memory_validity, memory_offset);
+  if (pool_size < min_pool_size || !validateLayout())
+    throw std::runtime_error("Planned layout is not feasible");
+
+  return double(pool_size) / double(min_pool_size);
```

Comment:
throw if min_pool size is zero? (more for svace...)

Reply:
this cant be zero -
1. if it is zero, its calculated at line 65.
2. this calculation can't be zero as memory_size is non empty, and memory pool does not accept requests of zero size.

Reply:
Alright :)